File: audio.r
library(seewave)
library(tuneR)

RATE = 44100
LEN_SECONDS = 0.39 # max length of each audio clip in seconds.
LEN_SAMPLES = LEN_SECONDS * RATE
N = 384 - 11

# spectrogram bands [0, C_3, C_4, C_5, 100]
# define: 
#   "low" as below C_3, 
#   "med" as C_3 to C_4 (middle C),
#   "high" as above C_4
MIDDLE_C = 261.63 / 1000 # kHz
BANDS = c(0, MIDDLE_C / 2, MIDDLE_C, 1000)

periods = data.frame(rbind(
    c("kato", "attack", 0.05, 0.11),
    c("kato", "sustain", 0.17, 0.24),
    c("kato", "decay", 0.30, LEN_SECONDS),
    c("maui", "attack", 0.04, 0.10),
    c("maui", "sustain", 0.12, 0.16),
    c("maui", "decay", 0.21, LEN_SECONDS),
    c("practice", "attack", 0.03, 0.08),
    c("practice", "sustain", 0.11, 0.17),
    c("practice", "decay", 0.29, LEN_SECONDS)), stringsAsFactors=F)
colnames(periods) = c("drum", "type", "start", "end")
periods$start = as.numeric(periods$start) * RATE
periods$end = as.numeric(periods$end) * RATE

players = drums = woods = angles = volumes = NULL # indices
extremes = attacks = sustains = decays = NULL # amplitude envelope features
lows = meds = highs = NULL # spectrogram features
row = 1
for (p in c("player1", "player2", "player3", "player4")) {
    for (d in c("kato", "maui", "practice")) {
        for (w in c("oak", "maple")) {
            for (a in c("1", "2", "3", "4")) {
                for (v in c("a", "b", "c", "d")) {
                    path = paste("wave_hits/", p, "_", d, "_", w, "_", a, v, ".wav", sep="")
                    if (!file.exists(path)) {
                        cat("  skipping ", path, "\n")
                        next
                    }
                    # record indices
                    players[row] <- p
                    drums[row] <- d
                    woods[row] <- w
                    angles[row] <- a
                    volumes[row] <- v

                    # find the amplitude envelope features
                    wave = readWave(path, to=LEN_SAMPLES)
                    envelope = env(wave, plot=F)
                    extremes[row] = quantile(envelope, 0.98)[[1]]

                    envelope_mean <- function(for_type) {
                        start = subset(periods, drum==d & type==for_type, start)[[1]]
                        end = subset(periods, drum==d & type==for_type, end)[[1]]
                        return(mean(envelope[start:end]))
                    }
                    attacks[row] = envelope_mean("attack")
                    sustains[row] = envelope_mean("sustain")
                    decays[row] = envelope_mean("decay")

                    ratios = fbands(spec(wave, plot=F), bands=BANDS, plot=F)[,2]
                    stopifnot(length(ratios) == 3)

                    lows[row] = ratios[1]
                    meds[row] = ratios[2]
                    highs[row] = ratios[3]

                    row <- row + 1
                }
            }
        }
    }
}

# Control for extremes separately for each drum
# Use angles == 2 as the reference
is_kato = ifelse(drums == "kato", 1, 0)
is_maui = ifelse(drums == "maui", 1, 0)
is_practice = ifelse(drums == "practice", 1, 0)
angle_1 = ifelse(angles == "1", 1, 0)
angle_3 = ifelse(angles == "3", 1, 0)
angle_4 = ifelse(angles == "4", 1, 0)

# TODO: upgrade these to pictures??
# print("Extremes vs angles:")
# model <- lm(extremes ~ angle_1 + angle_3 + angle_4)
# print(summary(model))

# print("Attack envelope magnitude, basic controls only:")
# model <- lm(attacks ~ angle_1 + angle_3 + angle_4
#        + is_kato + is_maui + is_practice - 1)
# print(summary(model))

# print("Sustain envelope magnitude, no controls:")
# model <- lm(sustains ~ angle_1 + angle_3 + angle_4)
# print(summary(model))

# print("Decay envelope magnitude, no controls:")
# model <- lm(decays ~ angle_1 + angle_3 + angle_4)
# print(summary(model))
 
print("Attack envelope magnitude, controlling for volume:")
model <- lm(attacks ~ drums + drums : poly(extremes, 3) 
                    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))
 
print("Sustain envelope magnitude, controlling for volume:")
model <- lm(sustains ~ drums + drums : poly(extremes, 3) 
                    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))

print("Decay envelope magnitude, controlling for volume:")
model <- lm(decays ~ drums + drums : poly(extremes, 3) 
                    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))

print("Lows by angle:")
model <- lm(lows ~ drums + drums : poly(extremes, 3) 
    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))

print("Mediums by angle:")
model <- lm(meds ~ drums + drums : poly(extremes, 3) 
    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))

print("Highs by angle:")
model <- lm(highs ~ drums + drums : poly(extremes, 3) 
    + angle_1 + angle_3 + angle_4 - 1)
print(summary(model))

# presumably suppresses everything except the mid-range
print("Zoom in on mediums for angle 4:")
model <- lm(meds ~ drums + drums : poly(extremes, 3) 
    + drums:angle_4 - 1)
print(summary(model))

File: audio.py
import numpy as np
import librosa as lb
import librosa.display
import matplotlib.pyplot as plt
from os.path import exists
from sklearn.linear_model import LinearRegression

players = ["player1", "player2", "player3", "player4"]
drums = ["kato", "maui", "practice"]
woods = ["oak", "maple"]
angles = ["1", "2", "3", "4"]
volumes = ["a", "b", "c", "d"]

LEN_SECONDS = 0.39 # max length of each audio clip in seconds.  This was chosen
# to get a reasonable sustain measurement on all drums, while only
# needing to throw away 4 data points for being to short.

periods_by_drum = {
    "kato":
        [   (0.05, 0.11),
            (0.17, 0.24),
            (0.30, LEN_SECONDS) ],
    "maui":
        [   (0.04, 0.10),
            (0.12, 0.16),
            (0.21, LEN_SECONDS) ],
    "practice":
        [   (0.03, 0.08),
            (0.11, 0.17),
            (0.29, LEN_SECONDS) ],
}
RATE = 22050 # sample rate in frames per second
FRAMES = int(LEN_SECONDS * RATE)

aud_list = np.zeros((4, 3, 2, 4, 4), dtype=object)

auds = np.zeros((4, 3, 2, 4, 4, FRAMES), dtype=np.float32)
ok = np.zeros((4, 3, 2, 4, 4), dtype=bool)

if exists("audio_matrix.npy"):
    print "loading saved audio matrix..."
    auds = np.load("audio_matrix.npy")
else:
    print "loading audio matrix from file..."
    # load the audio files by path
    for p, player in enumerate(players):
        for d, drum in enumerate(drums):
            for w, wood in enumerate(woods):
                for a, angle in enumerate(angles):
                    for v, volume in enumerate(volumes):
                        path = "hits/%s_%s_%s_%s%s.m4a" % (player, drum, wood, angle, volume)
                        if not exists(path):
                            print path, "doesn't exist; skipping"
                            continue
                        raw_aud, rate = lb.load(path)
                        assert rate == RATE
                        if len(raw_aud) < FRAMES:
                            print path, "too short; skipping"
                            continue
                        if max(raw_aud) < 0.02:
                            print path, "inaudibly soft; skipping"
                            continue
                        ok[p, d, w, a, v] = True
                        auds[p, d, w, a, v, 0:FRAMES] = raw_aud[0:FRAMES]
    np.save("audio_matrix.npy", auds)

for d, drum in enumerate(drums):
    print drum, ":"

    # for this drum, predict attack amplitude from max amplitude & max amplitude squared
    amps = np.abs(auds[:, d, :, :, :, :])
    oks = ok[:, d, :, :, :]
    extremes = np.percentile(amps, 98, axis = -1)[oks].reshape(-1, 1)
    X = np.hstack([extremes, extremes**2])

    start = int(periods_by_drum[drum][0][0] * rate)
    end = int(periods_by_drum[drum][0][1] * rate)
    attack_means = np.mean(amps[:, :, :, :, start:end], axis = -1)

    start = int(periods_by_drum[drum][1][0] * rate)
    end = int(periods_by_drum[drum][1][1] * rate)
    sustain_means = np.mean(amps[:, :, :, :, start:end], axis = -1)

    start = int(periods_by_drum[drum][2][0] * rate)
    end = int(periods_by_drum[drum][2][1] * rate)
    decay_means = np.mean(amps[:, :, :, :, start:end], axis = -1)

    y = attack_means[oks].reshape(-1, 1)
    attack_model = LinearRegression()
    attack_model.fit(X, y)
    print "\tAttack regression score from 2nd order extremes:", attack_model.score(X, y)
    attack_res = y - attack_model.predict(X)

    y = sustain_means[oks].reshape(-1, 1)
    sustain_model = LinearRegression()
    sustain_model.fit(X, y)
    print "\tSustain regression score from 2nd order extremes:", sustain_model.score(X, y)
    sustain_res = y - sustain_model.predict(X)

    y = decay_means[oks].reshape(-1, 1)
    decay_model = LinearRegression()
    decay_model.fit(X, y)
    print "\tDecay regression score from 2nd order extremes:", decay_model.score(X, y)
    decay_res = y - decay_model.predict(X)

    # for v, volume in enumerate(volumes):
    #     print "  ", volume, ":"

    #     amps = np.abs(auds[:, d, :, :, v, :])
    #     norm = np.max(amps, axis=-1)

        
    #     attack_props = np.mean(amps[:, :, :, start:end], axis=-1) / norm
    #     print "  avg attack prop:", np.mean(attack_props[ok[:,d,:,:,v]])
    #     # attack_avg = np.mean(np.abs(auds[p, d, :, :, :, start : end]))


    
    #     # sustain_avg = np.mean(np.abs(auds[p, d, :, :, :, start : end]))
    #     sustain_props = np.mean(amps[:, :, :, start:end], axis=-1) / norm
    #     print "  avg sustain prop:", np.mean(sustain_props[ok[:,d,:,:,v]])

       
    #     # decay_avg = np.mean(np.abs(auds[p, d, :, :, :, start : end]))
    #     decay_props = np.mean(amps[:, :, :, start:end], axis=-1) / norm
    #     print "  avg decay prop:", np.mean(decay_props[ok[:,d,:,:,v]])

    #     # m = np.max(np.abs(auds[p, d, :, :, :, :]))
    #     # print "    attack prop: %.4f" % (attack_avg / float(m))
    #     # print "    sustain prop: %.4f" % (sustain_avg / float(m))
    #     # print "    decay prop: %.4f" % (decay_avg / float(m))


# well... you could try to regress on volume and take the residuals
# also try avg volume
# expect +/- 0.02 ratio from player
# expect +/- 0.03 ratio from volume
